src/main/scala/sigmastate/lang/SigmaParser.scala

```diff
@@ -25,9 +28,9 @@ object SigmaParser extends Exprs with Types with Core {
 //    P( (Id | `this`).! ~ LambdaDef ).map { case (name, lam) => builder.mkVal(name, NoType, lam) }
 //  }
 
-  val ValVarDef = P( BindPattern/*.rep(1, ",".~/)*/ ~ (`:` ~/ Type).? ~ (`=` ~/ FreeCtx.Expr) ).map {
-    case (Ident(n,_), t, body) => builder.mkVal(n, t.getOrElse(NoType), body)
-    case (pat,_,_) => error(s"Only single name patterns supported but was $pat")
+  val ValVarDef = P( Index ~ BindPattern/*.rep(1, ",".~/)*/ ~ (`:` ~/ Type).? ~ (`=` ~/ FreeCtx.Expr) ).map {
+    case (index, Ident(n,_), t, body) => builder.mkVal(n, t.getOrElse(NoType), body, srcCtx(index))
```

Comment:
@greenhat In the suggested description of sourceContext above I mention  to use dynamic var.
Here it may look like
`parser.atSourcePos(index) { builder.mkVal(n, t.getOrElse(NoType), body) }`
Here `atSourcePos` some method you can add in the Parser to incapsulate dynamic variable machinery.

Reply:
Thanks. Done.